src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -491,195 +517,217 @@ protected override void OnConsumedBytes(int count)
                             break;
                         }
-                        else if (fin)
+                        else if (result.IsCompleted)
                         {
                             _context.RejectRequest(RequestRejectionReason.ChunkedRequestIncomplete);
                         }
-
-                        await _input;
                     }
                 }
 
                 // Chunks finished, parse trailers
                 while (_mode == Mode.Trailer)
                 {
-                    var fin = _input.CheckFinOrThrow();
+                    var result = await _input.ReadAsyncDispatched();
+                    var buffer = result.Buffer;
+                    var consumed = default(ReadCursor);
+                    var examined = default(ReadCursor);
 
-                    ParseChunkedTrailer();
+                    try
+                    {
+                        ParseChunkedTrailer(buffer, out consumed, out examined);
+                    }
+                    finally
+                    {
+                        _input.Advance(consumed, examined);
+                    }
 
                     if (_mode != Mode.Trailer)
                     {
                         break;
                     }
-                    else if (fin)
+                    else if (result.IsCompleted)
                     {
                         _context.RejectRequest(RequestRejectionReason.ChunkedRequestIncomplete);
                     }
 
-                    await _input;
                 }
 
                 if (_mode == Mode.TrailerHeaders)
                 {
-                    while (!_context.TakeMessageHeaders(_input, _requestHeaders))
+                    while (true)
                     {
-                        if (_input.CheckFinOrThrow())
+                        var result = await _input.ReadAsyncDispatched();
+                        var buffer = result.Buffer;
+
+                        if (buffer.IsEmpty && result.IsCompleted)
                         {
-                            if (_context.TakeMessageHeaders(_input, _requestHeaders))
-                            {
-                                break;
-                            }
-                            else
-                            {
-                                _context.RejectRequest(RequestRejectionReason.ChunkedRequestIncomplete);
-                            }
+                            _context.RejectRequest(RequestRejectionReason.ChunkedRequestIncomplete);
                         }
 
-                        await _input;
-                    }
+                        var consumed = default(ReadCursor);
+                        var examined = default(ReadCursor);
 
+                        bool takeMessageHeaders;
+                        try
+                        {
+                            takeMessageHeaders = _context.TakeMessageHeaders(buffer, _requestHeaders, out consumed, out examined);
```

Comment:
Why use a bool? Couldn't if put the `if (...) break;` here since you advance in a finally anyway?